### PR TITLE
getjson: add safe_format for None

### PR DESCRIPTION
**Pull request by lasers**
We like the format. Uses `self.py3.safe_format(self.format)`. Hassle-free PR. 🍉

### CODE REVIEW

**Comment by tobes** [comment]
I like this idea 👍

**Comment by tobes** [comment]
@lasers can you break this into 2 PRs one just with the doc changes and the other with the code ones.  I cannot review this.  It was really nice when I looked yesterday but you have added all the doc changes so now it is a real pain to review.  Please avoid reformatting docstrings etc just because you can, unless it is the only thing in the PR

**Comment by lasers** [comment]
@tobes Yesterday was nice and simple because of one-line change. Since we're passing `safe_format` either way, I go ahead and broom away unnecessary bread crumbs with `status` code. Here's two commits if this is okay.

**Comment by tobes** [comment]
better thanks

File: py3status/modules/getjson.py [hunk 1]
```diff
@@ -94,28 +92,23 @@ def _query_url(self):
         """
         """
         try:
-            resp = self.py3.request(self.url, timeout=self.timeout)
-            status = resp.status_code == 200
-            resp = resp.json()
+            resp = self.py3.request(self.url, timeout=self.timeout).json()
```
Comment (tobes) [inline py3status/modules/getjson.py]: It is possible that `.json` fails here eg set `url = 'google.co.uk'`

I think really the `py3.request` should handle this better and provide an exception that would be caught by the general `py3.RequestException`  but could also be caught explicitly something like `py3.RequestInvalidJSON`

File: py3status/modules/getjson.py [hunk 2]
```diff
@@ -68,28 +65,17 @@ class Py3status:
...
-            'cached_until': self.py3.time_in(self.cache_timeout),
-        }
-
         try:
-            resp = self.py3.request(self.url, timeout=self.timeout)
-            status = resp.status_code == 200
-            resp = resp.json()
+            getjson = self.py3.request(self.url, timeout=self.timeout).json()
+            getjson = self.py3.flatten_dict(
+                getjson, delimiter=self.delimiter, intermediates=True)
```
Comment (tobes) [inline py3status/modules/getjson.py]: imho this is ugly because you are between two approaches
```# I like this
function(
    params1, param2, param3
)

# sometimes this is great too
function(params1,
         param2,
         param3)


# this is bastard child that should not live
function(
    params1, param2, param3)
```

File: py3status/modules/getjson.py [hunk 3]
```diff
@@ -68,28 +65,17 @@ class Py3status:
...
-            response['full_text'] = self.py3.safe_format(
-                self.format,
-                self.py3.flatten_dict(
-                    resp, delimiter=self.delimiter, intermediates=True
-                )
-            )
-        else:
-            response['full_text'] = ''
-        return response
+            getjson = None
```
Comment (tobes) [inline py3status/modules/getjson.py]: This is interesting....

say I have format `info: {moo}`  when no data what do we wish to show?

`<nothing>`  <- what we show currently

`info: {moo}`  <- after this PR placeholders show as 'missing'

`info: ` <- I like this we make all placeholders `''` so they get hidden

what do you think?

Comment (lasers) [inline py3status/modules/getjson.py]: This is a good scenario. We don't know what `{placeholder}` we can use with `getjson`.

How do you suggest we `''` the unknown placeholders... but still display the format? I'm guessing same for-loop `empty_status` in deadbeef.

If I use `format = "GETJSON: {moo}"`, should I get `GETJSON:` or `GETJSON: {moo}` or `''`?

`<nothing>` <- what we show currently -- Goes against my `GETJSON: ` format.
`GETJSON: {moo}` <- after this PR placeholders show as 'missing'.  Right. We should update our format to `GETJSON[: {moo}]` then... but this shows us that the URL is not available right now... or something unexpected happened.

`info:` <- I like this we make all placeholders '' so they get hidden`.... I think this may be harder to make things consistent between modules. Easier to just pass things as-is... if things may not be available sometimes, then we update format with your weaponized `the-formatter` gun.

I think using `[{placeholder}]` is nice, makes sense, hints people to learn a bit more about the formatter... than adding for-loop `empty_status` in all modules. Idk.

I think it's good to display `GETJSON: {moo} 1232.34`... Shows us that we need to fix `{moo}` placeholder somehow.... than `GETJSON: 1234.43` then people might not know about this broken/wrong input/typos placeholders. Again, IDK.

Comment (tobes) [inline py3status/modules/getjson.py]: > I think it's good to display GETJSON: {moo} 1232.34... Shows us that we need to fix {moo} placeholder somehow.... than GETJSON: 1234.43 then people might not know about this broken/wrong input/typos placeholders. 

Yeah this is important and we do this when we have data returned

> How do you suggest we '' the unknown placeholders... but still display the format? 

piss easy - the question is what is the best way to do this.

options:

* create our dict of all placeholders being `''`

* tell formatter we don't want empty ones shown.

* just decide that if we have no data that we don't show (eg `safe_format(format, None)` as opposed to `safe_format(format, {})` (here we show missing)

anyhow try to stick to single issues, so what behaviour we want, not other format ideas, otherwise it makes these discussions really hard to follow and we lose stuff.

Comment (lasers) [inline py3status/modules/getjson.py]: I think we should use `[{}]`. That's what the formatter is for. When I was fixing `gpmdp`, I had wrong placeholders in. I forget we need to use `{song_artist}` instead of `{artist}` and `{song_title}` instead of `{title}`.

Hiding placeholders right away could make easy fixes turning into a much harder fixes for everybody. We get users coming in, saying they tried something... and it didn't work... Oh it's because they put in wrong placeholders, deprecated placeholders, or typos in placeholders, etc.

P.S. Being a detective suck... Hunting down bugs with tools that does not work as expected because things get blocked, things get hidden, etc... It's not a bug,  it's a feature, they all said. 🤕

Comment (tobes) [inline py3status/modules/getjson.py]: does not compute 🤖 we are talking about the 'no data' situation.

Comment (lasers) [inline py3status/modules/getjson.py]: >create our dict of all placeholders being ''

If we went with this, compute now?

>tell formatter we don't want empty ones shown.

I don't know py3status at all... This sounds sensible, but I like `[{}]` better. 

>just decide that if we have no data that we don't show (eg safe_format(format, None) as opposed to safe_format(format, {}) (here we show missing)

Now we have two point of entries. We don't want that here anyway.

File: py3status/modules/getjson.py (hunk 2 above)
Comment (lasers) [inline py3status/modules/getjson.py]: Went with `# I like this`. Thx for the tip. I like bastard child. Less lines for things that we likely won't need to play with.

File: py3status/modules/getjson.py (hunk 3 above)
Comment (tobes) [inline py3status/modules/getjson.py]: > Now we have two point of entries. We don't want that here iirc. (OKAY, MAYBE NOT HERE... but I'm referring to simple if-else + double safe_format thing for the sake of making things Noice AKA None + nice.

could you translate this please

Comment (lasers) [inline py3status/modules/getjson.py]: From what I understand... I'm returning getjson value right away without any checking.

```
# Direct return
return {
    'cached_until': self.py3.time_in(self.cache_timeout),
    'full_text': self.py3.safe_format(self.format, getjson)
}
```
... TO...
```
# Quick check before returning
if getjson:
    getjson = self.py3.safe_format(self.format, getjson)
else:
    getjson = self.py3.safe_format(self.format, None)

return {
    'cached_until': self.py3.time_in(self.cache_timeout),
    'full_text': getjson
}
```
We have to check this for all modules that could come back None.
And maybe we have to check for (some) placeholders that could come back None.

This feels like we're making things too noice for unknown placeholders.

Need to play with this to see which feels right with `eth0` up/down.
For other known placeholders, it's not an issue... and it also tells us that some placeholders may have no values at some times. I think `[{}]`. Idk.
```
getjson{
    url = 'http://ip-api.com/json'
    format = 'GETJSON: [{lat}, ]{lon} {wrong}'
}
```

Comment (tobes) [inline py3status/modules/getjson.py]: we'd just do this
```
return {
    'cached_until': self.py3.time_in(self.cache_timeout),
    'full_text': self.py3.safe_format(self.format, getjson),
}
```

The idea is less code everywhere.

So really the question is all about behaviour of the formatter.  I think we need to keep how this module works close to the original.

Yeah we'd need to see how other modules are affected.

So you might want to just do `full_text = ''` for a quick merge and then look at that issue over a few weeks or so.

Comment (lasers) [inline py3status/modules/getjson.py]: Sure. Will `self.py3.safe_format(self.format, None)` work? If not, I'll `''` it.

I think it's hard to anticipate how formatter will react if I return `''`. Will that turn `[TEST: {empty}]` --> `TEST: `? Maybe not, but it feels like this way.

Comment (lasers) [inline py3status/modules/getjson.py]: `self.py3.safe_format(self.format, None)` is same as `''` for me here with this format... `        format = 'GETJSON: [{lat}, ]{lon} {wrong}'` No GETJSON!!! Merge.

File: py3status/modules/getjson.py [hunk 4]
```diff
@@ -68,28 +65,21 @@ class Py3status:
...
-            response['full_text'] = self.py3.safe_format(
-                self.format,
-                self.py3.flatten_dict(
-                    resp, delimiter=self.delimiter, intermediates=True
-                )
-            )
+            getjson = None
+
+        if getjson:
+            getjson = self.py3.safe_format(self.format, getjson)
```
Comment (tobes) [inline py3status/modules/getjson.py]: `full_text`?

Comment (lasers) [inline py3status/modules/getjson.py]: Line 77.

Comment (tobes) [inline py3status/modules/getjson.py]: yeah and 81 too....

Comment (lasers) [inline py3status/modules/getjson.py]: Can we keep `getjson`? `full_text` is bastard child here when we reuse it.

File: py3status/modules/getjson.py [hunk 5]
```diff
@@ -68,28 +65,21 @@ class Py3status:
...
         """
-        response = {
-            'cached_until': self.py3.time_in(self.cache_timeout),
-        }
-
         try:
-            resp = self.py3.request(self.url, timeout=self.timeout)
-            status = resp.status_code == 200
-            resp = resp.json()
+            full_text = self.py3.request(self.url, timeout=self.timeout).json()
```
Comment (tobes) [inline py3status/modules/getjson.py]: The thing is that this isn't the `full_text` it is the json output so calling it `json_data` would be a better name.

It is much better to use appropriate meaningful names that describe what they are.  Reusing a variable for different meanings is a nice way to make code harder to read, prone to bugs etc.

File: py3status/modules/getjson.py [hunk 6]
```diff
@@ -68,28 +65,21 @@ class Py3status:
...
-            response['full_text'] = self.py3.safe_format(
-                self.format,
-                self.py3.flatten_dict(
-                    resp, delimiter=self.delimiter, intermediates=True
-                )
-            )
+            full_text = None
+
+        if full_text:
+            full_text = self.py3.safe_format(self.format, full_text)
```
Comment (tobes) [inline py3status/modules/getjson.py]: this is the point where `full_text` as a variable name makes sense.

**Comment by ultrabug** [comment]
How are we doing here mates ? I see @lasers addressed the review, is that okay ?

**Comment by tobes** [comment]
I've merged this there, thanks @lasers
